From patchwork Thu Aug 23 00:09:04 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
X-Patchwork-Id: 10573309
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id DA6A0112E
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 23 Aug 2018 00:09:30 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id BF3D72BAF9
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 23 Aug 2018 00:09:30 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B15F32BB0C; Thu, 23 Aug 2018 00:09:30 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 3787F2BAF9
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 23 Aug 2018 00:09:30 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727355AbeHWDgZ (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 22 Aug 2018 23:36:25 -0400
Received: from mail-pg1-f194.google.com ([209.85.215.194]:45381 "EHLO
        mail-pg1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727162AbeHWDgZ (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 22 Aug 2018 23:36:25 -0400
Received: by mail-pg1-f194.google.com with SMTP id m4-v6so1132810pgv.12;
        Wed, 22 Aug 2018 17:09:26 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=LwPbEyvT2o6FZTO012dfdKiR5CcSuJwfe1nIN8Q6NNg=;
        b=lzIkUcKCz/GD0hLWmgFjMbrBltY+xtlcUNHRDK2WCBnd3Fex7ABqXkrydsPV1AzIwH
         CJiB+mJEjBoctOl2+17oq9R4tK6a9bGex1bRDmAalIEbXay4i27/Vhiw20yP9LqYDCp5
         hSir0StTSvF5QU0sIn0HgNYYcaJHiye1n0ZXJbEa37vAILkjuwDCX3ePDrRknBnPKEsn
         edTTsZL2orA28OU5kV4UC7rhTZDxSVK0pmY+RrfuVYub1Umtk2hLTWYhEFvIIW+5zBno
         PeBRKHkdD/GTNisQaQKwjwFovw2Q7+Vz1G6yEzWbKuL6qUdYuc1rUxgxD6PxzC0iH86a
         5Iug==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=LwPbEyvT2o6FZTO012dfdKiR5CcSuJwfe1nIN8Q6NNg=;
        b=fvWbenkMumAuXbp59x0Xcr1+7wRnjrWRubjkt3Y+Lxc7OptbIHPP/4YCdGj5RbsDfh
         Qovystro/QvRT3PCt44Q8kZNE3M0GcJ0axS0MKJkSVRK89OeKV7eEbQY9ef99xnp8ALo
         X0BFzqfv+kd91Cb+CbNPMf7uK3S3evvfEqacqrPEFyR/pQfvyiY0KpO1+0Po8xBYIbxb
         RpGEAEaT4QHDb2JLXEVnZ6uGboi2uGzcnsad1uCMIkNzMAy58zQ7wYFa5ImpqdoViZZR
         mxdBt/WznS+cD32TfZh5cs7UKmS3mfq48Wsonr0ZeebrKNuKUUFxqD+xsFM7nMqLb29S
         gvgw==
X-Gm-Message-State: AOUpUlHwZ3uDpSjJKmTDG1FCVj4oJmMLGpv8n7BTWy79VQ8MVKOFwtnD
        JvkuXVlYaFuC2fZlpzZLAwb2vgHc
X-Google-Smtp-Source: 
 AA+uWPzZa2ZnF4PYV4+WaQ27CSU9lZ51sM0e7k28Jf2xuJ3PQ+MenIwYx8e4M7lfk1zaJVz/MdGoNA==
X-Received: by 2002:a63:e457:: with SMTP id
 i23-v6mr53943454pgk.127.1534982966399;
        Wed, 22 Aug 2018 17:09:26 -0700 (PDT)
Received: from surajjs2.ozlabs.ibm.com ([122.99.82.10])
        by smtp.gmail.com with ESMTPSA id
 x24-v6sm3973467pfh.67.2018.08.22.17.09.24
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Wed, 22 Aug 2018 17:09:25 -0700 (PDT)
From: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
To: kvm-ppc@vger.kernel.org, kvm@vger.kernel.org
Cc: paulus@ozlabs.org, sjitindarsingh@gmail.com
Subject: [PATCH] KVM: PPC: Book3S HV: Set fault_dsisr on H_INST_STORAGE
 interrupts
Date: Thu, 23 Aug 2018 10:09:04 +1000
Message-Id: <20180823000904.25845-1-sjitindarsingh@gmail.com>
X-Mailer: git-send-email 2.13.6
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

When a page fault is handled the bits in fault_dsisr are used to
determine the cause of the fault. In the H_DATA_STORAGE interrupt
case these bits come from the HDSISR register and are stored in
fault_dsisr on vm exit. However for a H_INST_STORAGE interrupt
fault_dsisr is set to zero.

Currently we handle the page fault anyway, even if fault_dsisr is set
to zero. However it means we can never update the RC bits of a pte on
an instruction fault since the corresponding bit to indicate this is
needed will never be set.

On a H_INST_STORAGE interrupt the fault bits are in HSRR1 which is
stored in the guest_msr on vm exit. So to correct this we mask the
appropriate bits from the guest_msr and set it in fault_dsisr.

Signed-off-by: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
---
 arch/powerpc/kvm/book3s_hv.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index ee4a8854985e..c5d29adec19a 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -1190,7 +1190,8 @@ static int kvmppc_handle_exit_hv(struct kvm_run *run, struct kvm_vcpu *vcpu,
 		break;
 	case BOOK3S_INTERRUPT_H_INST_STORAGE:
 		vcpu->arch.fault_dar = kvmppc_get_pc(vcpu);
-		vcpu->arch.fault_dsisr = 0;
+		vcpu->arch.fault_dsisr = kvmppc_get_msr(vcpu) &
+					 DSISR_SRR1_MATCH_64S;
 		r = RESUME_PAGE_FAULT;
 		break;
 	/*
